{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "G04iVpOTvZI7"
      },
      "source": [
        "# **MergingShitâ„¢ V2 based on ColabADMIXTOOLS Version 4.8a**\n",
        "Simplified, use it to merge shit. Done by pepsimanfire (Discord) but all credits to Florio who did the OG collab. florus14 his Discord. There's no point in merging on RStudio or whatsoever, merging requires TOO MUCH computational power."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BtePdSNjCQ1U"
      },
      "source": [
        "# **[1] Install Software**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "w4Gy9JnoyqEE"
      },
      "source": [
        "You can run with CPU if merging with the HO, but only with TPU 2.8 Runtime if 1240k. Pro tip: if you run out of TPU time, you can always use a different Google account, or create a new one."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QTN1OE0F_Kqj",
        "outputId": "4eb7f2e9-ed15-4652-9a0c-c79e982f1dc7"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "from subprocess import run, PIPE\n",
        "\n",
        "# @markdown **Select your preferences for the setup process**\n",
        "\n",
        "# @markdown Do you want to install ADMIXTOOLS?\n",
        "install_admixtools = True  # @param {type:\"boolean\"}\n",
        "\n",
        "# Function to check if a command exists\n",
        "def command_exists(command):\n",
        "    return run(['which', command], stdout=PIPE, stderr=PIPE).returncode == 0\n",
        "\n",
        "# Install R if not already installed\n",
        "if not command_exists('R'):\n",
        "    print(\"Installing R...\")\n",
        "    !sudo apt-get update -y\n",
        "    !sudo apt-get install -y r-base\n",
        "else:\n",
        "    print(\"R is already installed.\")\n",
        "\n",
        "# Install rpy2 if not already installed\n",
        "try:\n",
        "    import rpy2\n",
        "    print(\"rpy2 is already installed.\")\n",
        "except ImportError:\n",
        "    print(\"Installing rpy2...\")\n",
        "    !pip install rpy2\n",
        "\n",
        "# Load the rpy2 extension for Python\n",
        "%load_ext rpy2.ipython\n",
        "\n",
        "# Check for PLINK\n",
        "if not os.path.exists('/content/plink'):\n",
        "    print(\"Installing PLINK (latest stable version)...\")\n",
        "    !wget https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20241022.zip -O plink.zip\n",
        "    !unzip -o plink.zip -d /content/\n",
        "    os.remove(\"plink.zip\")\n",
        "    print(\"PLINK installation complete!\")\n",
        "else:\n",
        "    print(\"PLINK is already installed.\")\n",
        "\n",
        "# Check and install ADMIXTOOLS\n",
        "if install_admixtools:\n",
        "    if not os.path.exists('/content/AdmixTools'):\n",
        "        print(\"Installing ADMIXTOOLS...\")\n",
        "        !apt-get update\n",
        "        !apt-get install -y build-essential libgsl-dev libopenblas-dev gfortran liblapacke-dev \\\n",
        "                           libssl-dev libffi-dev libncurses5-dev zlib1g zlib1g-dev \\\n",
        "                           libreadline-dev libbz2-dev libsqlite3-dev\n",
        "        !rm -rf AdmixTools\n",
        "        !git clone https://github.com/DReichLab/AdmixTools.git\n",
        "        %cd AdmixTools/src\n",
        "        !sed -i \"s/-Wimplicit/-Wimplicit -fcommon/\" Makefile\n",
        "        !make clobber\n",
        "        !make all LDLIBS=\"-llapacke -llapack -lgsl -lopenblas -lm -lnick\"\n",
        "        !make install\n",
        "        %cd /content/\n",
        "        print(\"ADMIXTOOLS installation complete!\")\n",
        "    else:\n",
        "        print(\"ADMIXTOOLS is already installed.\")\n",
        "\n",
        "\n",
        "print(\"Setup complete!\")\n",
        "\n",
        "# Add this right after ADMIXTOOLS installation\n",
        "os.environ['PATH'] += os.pathsep + '/content/AdmixTools/bin'\n",
        "print(\"ADMIXTOOLS binaries added to PATH\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "irfC9KGv0dWw"
      },
      "source": [
        "# **[2] Your DNA File Upload**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 88
        },
        "id": "1OTzXLCGh32l",
        "outputId": "9925558a-82b5-40e6-f931-d6e111e011a2"
      },
      "outputs": [],
      "source": [
        "# @title Upload 23andMe Raw Data\n",
        "from google.colab import files\n",
        "\n",
        "uploaded = files.upload()\n",
        "\n",
        "# Assuming the user will upload only one genome file, you can check if a file has been uploaded successfully.\n",
        "if uploaded:\n",
        "    filename = next(iter(uploaded))\n",
        "    print(f\"File '{filename}' uploaded successfully.\")\n",
        "else:\n",
        "    print(\"No file uploaded.\")\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "oSmLy92EKmAT"
      },
      "source": [
        "**23andMe formatted data works best for conversion. Please explore [DNA Kit\n",
        "Studio](https://www.dnagenics.com/dna-kit-studio/) on how to make this conversion possible. Some users have also found success with [this](https://tendna.com/en/dna-converter) converter as well.**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zIHcUfjY1DV7"
      },
      "source": [
        "# **[3] Dataset Selection and snplist download**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "l9Am_c6ONKLn"
      },
      "source": [
        "**The 1240K has more coverage than the 1240K+HO. However, the 1240K+HO has more modern samples. You can download one after the other if you would like to explore both datasets.**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "e-9bhlIJvT1o",
        "outputId": "34aadf09-7646-430a-bcd5-a28706657cdf"
      },
      "outputs": [],
      "source": [
        "# Clone the repo into /content/snplist\n",
        "!git clone https://github.com/astralprojectionfont/snplist.git\n",
        "\n",
        "# Verify the download\n",
        "!ls -la snplist/\n",
        "\n",
        "# @title Download 1240K or HO Datasets (AADR v62)\n",
        "# @markdown Select the dataset you want to download:\n",
        "\n",
        "dataset_choice = \"1240K\" #@param [\"1240K\", \"1240K+HO\"]\n",
        "\n",
        "# Define the URLs and corresponding filenames for each dataset\n",
        "dataset_files = {\n",
        "    \"1240K\": [\n",
        "        (\"https://dataverse.harvard.edu/api/access/datafile/10537413\", \"aadr_v62.0_1240K_public.anno\"),\n",
        "        (\"https://dataverse.harvard.edu/api/access/datafile/10537126\", \"aadr_v62.0_1240K_public.geno\"),\n",
        "        (\"https://dataverse.harvard.edu/api/access/datafile/10537414\", \"aadr_v62.0_1240K_public.ind\"),\n",
        "        (\"https://dataverse.harvard.edu/api/access/datafile/10537415\", \"aadr_v62.0_1240K_public.snp\"),\n",
        "    ],\n",
        "    \"1240K+HO\": [\n",
        "        (\"https://dataverse.harvard.edu/api/access/datafile/10537417\", \"aadr_v62.0_HO_public.anno\"),\n",
        "        (\"https://dataverse.harvard.edu/api/access/datafile/10537419\", \"aadr_v62.0_HO_public.geno\"),\n",
        "        (\"https://dataverse.harvard.edu/api/access/datafile/10537420\", \"aadr_v62.0_HO_public.ind\"),\n",
        "        (\"https://dataverse.harvard.edu/api/access/datafile/10537421\", \"aadr_v62.0_HO_public.snp\"),\n",
        "    ]\n",
        "}\n",
        "\n",
        "selected_files = dataset_files[dataset_choice]\n",
        "\n",
        "print(f\"Downloading {dataset_choice} dataset files...\")\n",
        "for url, filename in selected_files:\n",
        "    !wget --no-check-certificate {url} -O /content/{filename}\n",
        "\n",
        "print(\"Download complete!\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9DQIkKnjPSIl"
      },
      "source": [
        "# **[4] Merging Yourself with the AADR**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6jKui3wT9vvk"
      },
      "source": [
        "The code will ask you for what dataset you want to use, if you want to trim using the snp list or not, the sex the samples are (1=Male 2=Female 0=Unknown) and for a name for the samples in the ind file. For the rest, it auto detects txts as samples. The more stuff you add, the more time it will take of course. I think it detects up to 10 samples, which I guess you will never use, that'd be TOO much time.\n",
        "Many times, when merging samples from different companies, or just randomly, the script will fail, likely due to an issue with the samples and plink. I advise you to check the plink log, and to either try to remove the problematic raw file, or just merge one by one. In other cases, the conversion of the raw to the 23andme format can cause problems, it's kind of random really.\n",
        "It's not possible to merge everything."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "id": "4cDNUPDEU0O6",
        "outputId": "c1d90878-2821-4522-c500-dcaa0bf04e2d"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Select dataset (1 for HO, 2 for 1240K): 2\n",
            "Enter sex for /content/zohan23andme.txt (1 for male, 2 for female, 0 for unknown): 1\n",
            "Enter name for /content/zohan23andme.txt (to be displayed in .ind file): Zohan.DG\n",
            "Enter output base prefix: 4Zohan\n",
            "Do you want to trim using a SNP list? (yes/no): yes\n",
            "Swap memory created successfully.\n",
            "Processing /content/zohan23andme.txt...\n",
            ".fam file updated: /content/4Zohan_1_trimmed.fam\n",
            "Merging PLINK files...\n",
            "Merged PLINK file created: 4Zohan_merged\n",
            "convertf parameter file created successfully.\n",
            "Running mergeit...\n",
            "parameter file: /content/mergeit_param.par\n",
            "geno1: aadr_v62.0_1240K_public.geno\n",
            "snp1: aadr_v62.0_1240K_public.snp\n",
            "ind1: aadr_v62.0_1240K_public.ind\n",
            "geno2: 4Zohan_converted.geno\n",
            "snp2: 4Zohan_converted.snp\n",
            "ind2: 4Zohan_converted.ind\n",
            "genooutfilename: 4Zohan_merged.geno\n",
            "snpoutfilename: 4Zohan_merged.snp\n",
            "indoutfilename: 4Zohan_merged.ind\n",
            "testmismatch: NO\n",
            "numthreads: 2\n",
            "## mergeit version: 3130\n",
            "allele funny: rs2821054 G T X X\n",
            "allele funny: rs4845877 T C X X\n",
            "allele funny: rs2256787 A C X X\n",
            "allele funny: rs6698850 C T X X\n",
            "allele funny: rs1168029 G A X X\n",
            "allele funny: rs12408354 A C X X\n",
            "allele funny: rs11163136 G A X X\n",
            "allele funny: rs9782928 C T X X\n",
            "allele funny: rs858542 G T X X\n",
            "allele funny: rs35266764 T C X X\n",
            "allele funny: rs639524 A G X X\n",
            "allele funny: rs1367068 A G X X\n",
            "allele funny: rs5050 T G X X\n",
            "allele funny: rs350729 T G X X\n",
            "allele funny: rs17029711 C A X X\n",
            "allele funny: rs7572857 G A X X\n",
            "allele funny: rs6732651 T C X X\n",
            "allele funny: rs1471641 C T X X\n",
            "allele funny: rs2215599 G A X X\n",
            "allele funny: rs2728041 C T X X\n",
            "allele funny: rs9836291 G A X X\n",
            "allele funny: rs7629936 G A X X\n",
            "allele funny: rs1461616 C A X X\n",
            "allele funny: rs9844302 C T X X\n",
            "allele funny: rs7640527 C T X X\n",
            "allele funny: rs1188588 C T X X\n",
            "allele funny: rs6446395 G A X X\n",
            "allele funny: rs3846423 A G X X\n",
            "allele funny: rs2518649 T G X X\n",
            "allele funny: rs1966546 T C X X\n",
            "allele funny: rs1585959 T C X X\n",
            "allele funny: rs1126672 G A X X\n",
            "allele funny: rs1031691 T C X X\n",
            "allele funny: rs10857173 G A X X\n",
            "allele funny: rs1320100 A G X X\n",
            "allele funny: rs30774 T C X X\n",
            "allele funny: rs1435231 C A X X\n",
            "allele funny: rs4235652 C T X X\n",
            "allele funny: rs384267 G T X X\n",
            "allele funny: rs4631137 G A X X\n",
            "allele funny: rs700585 C T X X\n",
            "allele funny: rs622266 T C X X\n",
            "allele funny: rs12656675 A G X X\n",
            "allele funny: rs1382713 G T X X\n",
            "allele funny: rs1422993 A C X X\n",
            "allele funny: rs4557511 C T X X\n",
            "allele funny: rs1495978 G A X X\n",
            "allele funny: rs6931514 A G X X\n",
            "allele funny: rs11961941 G T X X\n",
            "allele funny: rs2844786 C A X X\n",
            "allele funny: rs3132645 T C X X\n",
            "allele funny: rs34435337 G A X X\n",
            "allele funny: rs3778639 A G X X\n",
            "allele funny: rs2523647 A G X X\n",
            "allele funny: rs15574 G A X X\n",
            "allele funny: rs2280774 G A X X\n",
            "allele funny: rs9268980 A G X X\n",
            "allele funny: rs9271255 C T X X\n",
            "allele funny: rs522308 C T X X\n",
            "allele funny: rs7756935 C A X X\n",
            "allele funny: rs973205 C A X X\n",
            "allele funny: rs16887931 T G X X\n",
            "allele funny: rs236874 A G X X\n",
            "allele funny: rs75484629 G A X X\n",
            "allele funny: rs10457410 C T X X\n",
            "allele funny: rs7745274 G A X X\n",
            "allele funny: rs631204 C A X X\n",
            "allele funny: rs2280559 T C X X\n",
            "allele funny: rs3012437 T C X X\n",
            "allele funny: rs1982157 A C X X\n",
            "allele funny: rs2526098 G A X X\n",
            "allele funny: rs6462167 G T X X\n",
            "allele funny: rs700750 C A X X\n",
            "allele funny: rs3757419 C T X X\n",
            "allele funny: rs2158146 T G X X\n",
            "allele funny: rs13229505 G T X X\n",
            "allele funny: rs9297145 C A X X\n",
            "allele funny: rs17152281 C A X X\n",
            "allele funny: rs12531078 C T X X\n",
            "allele funny: rs3735319 A G X X\n",
            "allele funny: rs6997305 G A X X\n",
            "allele funny: rs7822301 T G X X\n",
            "allele funny: rs8192306 A G X X\n",
            "allele funny: rs6996522 A C X X\n",
            "allele funny: rs6471757 G A X X\n",
            "allele funny: rs7824974 A G X X\n",
            "allele funny: rs10958215 G A X X\n",
            "allele funny: rs10815590 G A X X\n",
            "allele funny: rs10815964 G A X X\n",
            "allele funny: rs2153417 A G X X\n",
            "allele funny: rs2046841 G A X X\n",
            "allele funny: rs620090 G T X X\n",
            "allele funny: rs7856656 C T X X\n",
            "allele funny: rs11141915 A C X X\n",
            "allele funny: rs10820798 G A X X\n",
            "allele funny: rs10441737 C T X X\n",
            "allele funny: rs2900537 T C X X\n",
            "allele funny: rs2787334 T C X X\n",
            "allele funny: rs548403 A G X X\n",
            "allele funny: rs1124602 G A X X\n",
            "allele funny: rs4372050 T C X X\n",
            "allele funny: rs2808055 C T X X\n",
            "allele funny: rs4838541 C T X X\n",
            "allele funny: rs7923617 C A X X\n",
            "allele funny: rs527458 A G X X\n",
            "allele funny: rs1902681 T G X X\n",
            "allele funny: rs4353229 T C X X\n",
            "allele funny: rs841825 C T X X\n",
            "allele funny: rs4282943 T G X X\n",
            "allele funny: rs3793912 T G X X\n",
            "allele funny: rs7128291 G T X X\n",
            "allele funny: rs1994904 C A X X\n",
            "allele funny: rs12789246 G A X X\n",
            "allele funny: rs10838021 T C X X\n",
            "allele funny: rs12420625 T G X X\n",
            "allele funny: rs11236808 G A X X\n",
            "allele funny: rs12282752 T C X X\n",
            "allele funny: rs11220708 T C X X\n",
            "allele funny: rs7939387 T C X X\n",
            "allele funny: rs10744553 C T X X\n",
            "allele funny: rs2537760 T C X X\n",
            "allele funny: rs7301016 A G X X\n",
            "allele funny: rs12312033 A C X X\n",
            "allele funny: rs7313903 A C X X\n",
            "allele funny: rs10778536 G A X X\n",
            "allele funny: rs2268389 A G X X\n",
            "allele funny: rs645510 C T X X\n",
            "allele funny: rs1718120 T G X X\n",
            "allele funny: rs1843242 C T X X\n",
            "allele funny: rs10902503 T C X X\n",
            "allele funny: rs9526569 C T X X\n",
            "allele funny: rs3850071 A C X X\n",
            "allele funny: rs7984811 T C X X\n",
            "allele funny: rs9582230 T C X X\n",
            "allele funny: rs798880 G T X X\n",
            "allele funny: rs966918 C A X X\n",
            "allele funny: rs8012882 C A X X\n",
            "allele funny: rs4077378 T C X X\n",
            "allele funny: rs1811753 G A X X\n",
            "allele funny: rs602192 T C X X\n",
            "allele funny: rs7173128 G A X X\n",
            "allele funny: rs2289440 G A X X\n",
            "allele funny: rs17649409 T C X X\n",
            "allele funny: rs56257742 C T X X\n",
            "allele funny: rs4843165 C T X X\n",
            "allele funny: rs2019604 T G X X\n",
            "allele funny: rs4074549 T C X X\n",
            "allele funny: rs4073769 A G X X\n",
            "allele funny: rs16957494 A G X X\n",
            "allele funny: rs2277672 G A X X\n",
            "allele funny: rs8080966 C T X X\n",
            "allele funny: rs1476811 C T X X\n",
            "allele funny: rs445507 T C X X\n",
            "allele funny: rs632491 C T X X\n",
            "allele funny: rs1835653 G A X X\n",
            "allele funny: rs9630799 C A X X\n",
            "allele funny: rs1059442 T C X X\n",
            "allele funny: rs9966498 G A X X\n",
            "allele funny: rs2602699 G A X X\n",
            "allele funny: rs10411936 A G X X\n",
            "allele funny: rs12609641 C T X X\n",
            "allele funny: rs13345681 A C X X\n",
            "allele funny: rs7245984 A G X X\n",
            "allele funny: rs3960965 C A X X\n",
            "allele funny: rs625372 T C X X\n",
            "allele funny: rs12624732 C T X X\n",
            "allele funny: rs7273626 A G X X\n",
            "allele funny: rs74695908 C T X X\n",
            "allele funny: rs6027060 C T X X\n",
            "allele funny: rs2823572 A G X X\n",
            "allele funny: rs2827938 T C X X\n",
            "allele funny: rs5761894 T C X X\n",
            "allele funny: rs12393627 A G X X\n",
            "allele funny: rs6641821 T C X X\n",
            "allele funny: rs5915682 G A X X\n",
            "allele funny: rs5962096 G A X X\n",
            "allele funny: rs12008474 G A X X\n",
            "allele funny: rs6639766 T G X X\n",
            "allele funny: rs7888834 C T X X\n",
            "allele funny: rs7061250 A G X X\n",
            "allele funny: rs7889175 T C X X\n",
            "allele funny: rs4826852 T C X X\n",
            "allele funny: rs9887315 C T X X\n",
            "allele funny: rs5934384 A G X X\n",
            "allele funny: rs4830389 A G X X\n",
            "allele funny: rs881599 A G X X\n",
            "allele funny: rs2188932 T G X X\n",
            "allele funny: rs2015165 C T X X\n",
            "allele funny: rs7876907 G A X X\n",
            "allele funny: rs2188374 T C X X\n",
            "allele funny: rs7892146 G A X X\n",
            "allele funny: rs7877529 T C X X\n",
            "allele funny: rs5978463 T G X X\n",
            "allele funny: rs5979465 T G X X\n",
            "allele funny: rs17150 G A X X\n",
            "allele funny: rs1266348 G A X X\n",
            "allele funny: rs5935412 A G X X\n",
            "allele funny: rs5741880 G T X X\n",
            "allele funny: rs5935465 G A X X\n",
            "allele funny: rs11095614 G T X X\n",
            "allele funny: rs11095617 A G X X\n",
            "allele funny: rs2285633 T C X X\n",
            "allele funny: rs2375729 T G X X\n",
            "allele funny: rs1983287 T C X X\n",
            "allele funny: rs6632882 T C X X\n",
            "allele funny: rs12392874 C A X X\n",
            "allele funny: rs5955621 G A X X\n",
            "allele funny: rs6633185 C T X X\n",
            "allele funny: rs5955695 A G X X\n",
            "allele funny: rs7051930 A G X X\n",
            "allele funny: rs7471538 A G X X\n",
            "allele funny: rs1350837 T C X X\n",
            "allele funny: rs5990936 C T X X\n",
            "allele funny: rs4403524 G A X X\n",
            "allele funny: rs5925583 A C X X\n",
            "allele funny: rs5970629 G A X X\n",
            "allele funny: rs7059086 A C X X\n",
            "allele funny: rs5970867 A C X X\n",
            "allele funny: rs4828947 A C X X\n",
            "allele funny: rs12690411 T G X X\n",
            "allele funny: rs5925745 C T X X\n",
            "allele funny: rs5971088 A G X X\n",
            "allele funny: rs10126378 C T X X\n",
            "allele funny: rs2188615 T C X X\n",
            "allele funny: rs5944718 A G X X\n",
            "allele funny: rs4141087 T G X X\n",
            "allele funny: rs6526959 C A X X\n",
            "allele funny: rs950353 T C X X\n",
            "allele funny: rs4829182 A G X X\n",
            "allele funny: rs6631113 T C X X\n",
            "allele funny: rs2692986 T C X X\n",
            "allele funny: rs1718053 C T X X\n",
            "allele funny: rs6628685 A G X X\n",
            "allele funny: rs228342 G T X X\n",
            "allele funny: rs12845594 T C X X\n",
            "allele funny: rs6527246 G A X X\n",
            "allele funny: rs7883529 T C X X\n",
            "allele funny: rs2148103 A G X X\n",
            "allele funny: rs764070 G A X X\n",
            "allele funny: rs12009944 C A X X\n",
            "allele funny: rs5927172 A C X X\n",
            "allele funny: rs7060250 T C X X\n",
            "allele funny: rs5964227 A G X X\n",
            "allele funny: rs4827349 C A X X\n",
            "allele funny: rs5963533 G A X X\n",
            "allele funny: rs198785 C A X X\n",
            "allele funny: rs5917703 A G X X\n",
            "allele funny: rs5963687 A G X X\n",
            "allele funny: rs2961401 A G X X\n",
            "allele funny: rs5963841 T C X X\n",
            "allele funny: rs1144688 C A X X\n",
            "allele funny: rs7886668 A G X X\n",
            "allele funny: rs1339774 A C X X\n",
            "allele funny: rs17310797 T G X X\n",
            "allele funny: rs6609437 T C X X\n",
            "allele funny: rs5905620 A G X X\n",
            "allele funny: rs2480439 A G X X\n",
            "allele funny: rs5915139 A G X X\n",
            "allele funny: rs1332008 A G X X\n",
            "allele funny: rs5964688 T C X X\n",
            "allele funny: rs6525037 G T X X\n",
            "allele funny: rs938059 C A X X\n",
            "allele funny: rs1327484 G T X X\n",
            "allele funny: rs7886577 A G X X\n",
            "allele funny: rs1555776 T C X X\n",
            "allele funny: rs5936959 A G X X\n",
            "allele funny: rs17311899 C A X X\n",
            "allele funny: rs112005459 T G X X\n",
            "allele funny: rs7881752 C T X X\n",
            "allele funny: rs1597261 T G X X\n",
            "allele funny: rs4590575 G T X X\n",
            "allele funny: rs402851 T C X X\n",
            "allele funny: rs12012589 T G X X\n",
            "allele funny: rs5913393 C T X X\n",
            "allele funny: rs6622648 T G X X\n",
            "allele funny: rs5921922 G T X X\n",
            "allele funny: rs4828253 C T X X\n",
            "allele funny: rs5968918 A G X X\n",
            "allele funny: rs5922392 A C X X\n",
            "allele funny: rs116064975 T G X X\n",
            "allele funny: rs5924478 G A X X\n",
            "allele funny: rs5942373 A C X X\n",
            "allele funny: rs5941593 T C X X\n",
            "allele funny: rs2208238 C T X X\n",
            "allele funny: rs6619280 T C X X\n",
            "allele funny: rs5983147 G A X X\n",
            "allele funny: rs5983343 G A X X\n",
            "allele funny: rs6522759 G A X X\n",
            "allele funny: rs1413947 C T X X\n",
            "allele funny: rs1206625 A C X X\n",
            "allele funny: rs707286 G A X X\n",
            "allele funny: rs5921512 C T X X\n",
            "allele funny: rs7891591 T C X X\n",
            "allele funny: rs2157372 T C X X\n",
            "allele funny: rs17284928 A G X X\n",
            "allele funny: rs6621992 G T X X\n",
            "allele funny: rs5962673 T C X X\n",
            "allele funny: rs16985291 T C X X\n",
            "allele funny: rs1266720 G T X X\n",
            "allele funny: rs537111 C T X X\n",
            "allele funny: rs5942906 T C X X\n",
            "allele funny: rs4893416 G A X X\n",
            "allele funny: rs5982528 A C X X\n",
            "allele funny: rs3007200 A C X X\n",
            "allele funny: rs7891628 A C X X\n",
            "allele funny: rs6421119 T C X X\n",
            "allele funny: rs11260118 C T X X\n",
            "allele funny: rs5958727 T C X X\n",
            "allele funny: rs261724 G A X X\n",
            "allele funny: rs9988376 G T X X\n",
            "allele funny: rs5910573 T G X X\n",
            "allele funny: rs403558 C T X X\n",
            "allele funny: rs4825403 A G X X\n",
            "allele funny: rs5957365 T G X X\n",
            "allele funny: rs2196260 G A X X\n",
            "allele funny: rs651595 C T X X\n",
            "allele funny: rs6634627 C T X X\n",
            "allele funny: rs112124312 T G X X\n",
            "allele funny: rs1341670 G A X X\n",
            "allele funny: rs10856742 A C X X\n",
            "allele funny: rs5930669 C T X X\n",
            "allele funny: rs2151083 C T X X\n",
            "allele funny: rs742217 C T X X\n",
            "allele funny: rs5931320 C T X X\n",
            "allele funny: rs5931498 T C X X\n",
            "allele funny: rs660684 C T X X\n",
            "allele funny: rs5931627 A C X X\n",
            "allele funny: rs10126631 T C X X\n",
            "allele funny: rs12688185 G A X X\n",
            "allele funny: rs5907663 T C X X\n",
            "allele funny: rs6636325 G A X X\n",
            "allele funny: rs6528934 C T X X\n",
            "allele funny: rs10521811 A G X X\n",
            "allele funny: rs5907262 T C X X\n",
            "allele funny: rs991072 A C X X\n",
            "allele funny: rs5953805 C T X X\n",
            "allele funny: rs2903031 G T X X\n",
            "allele funny: rs5955025 A C X X\n",
            "allele funny: rs851075 A C X X\n",
            "allele funny: rs4825194 T C X X\n",
            "allele funny: rs1930676 A G X X\n",
            "allele funny: rs9698285 T C X X\n",
            "allele funny: rs5905024 C T X X\n",
            "allele funny: rs2742895 C T X X\n",
            "allele funny: rs12009290 A C X X\n",
            "allele funny: rs25704 C T X X\n",
            "allele funny: rs2392878 G A X X\n",
            "allele funny: rs5980623 C T X X\n",
            "allele funny: rs4844019 C T X X\n",
            "allele funny: rs7883741 T C X X\n",
            "allele funny: rs5924884 C T X X\n",
            "allele funny: rs2070097 A G X X\n",
            "allele funny: rs3903 G T X X\n",
            "allele funny: rs111927750 G T X X\n",
            "allele funny: rs9786021 C A X X\n",
            "allele funny: rs4568775 C A X X\n",
            "allele funny: rs7067348 T C X X\n",
            "allele funny: rs9786834 G A X X\n",
            "allele funny: rs13447347 G T X X\n",
            "allele funny: rs6530626 A C X X\n",
            "read 1073741824 bytes\n",
            "read 2147483648 bytes\n",
            "read 3221225472 bytes\n",
            "read 4294967296 bytes\n",
            "read 5368709120 bytes\n",
            "read 5435121304 bytes\n",
            "packed geno read OK\n",
            "end of inpack\n",
            "packed geno read OK\n",
            "end of inpack\n"
          ]
        }
      ],
      "source": [
        "import os\n",
        "import subprocess\n",
        "\n",
        "# Step 1: Select Dataset\n",
        "while True:\n",
        "    dataset_choice = input(\"Select dataset (1 for HO, 2 for 1240K): \").strip()\n",
        "    if dataset_choice in ['1', '2']:\n",
        "        dataset_name = \"HO\" if dataset_choice == '1' else \"1240K\"\n",
        "        break\n",
        "    print(\"Invalid choice. Please enter 1 for HO or 2 for 1240K.\")\n",
        "\n",
        "# Merge Dataset Parameters\n",
        "geno1_path = f\"aadr_v62.0_{dataset_name}_public.geno\"\n",
        "snp1_path = f\"aadr_v62.0_{dataset_name}_public.snp\"\n",
        "ind1_path = f\"aadr_v62.0_{dataset_name}_public.ind\"\n",
        "\n",
        "# Specify Inputs\n",
        "genome_filepaths = [f\"/content/{f}\" for f in os.listdir(\"/content/\") if f.endswith(\".txt\")]\n",
        "\n",
        "if not genome_filepaths:\n",
        "    raise ValueError(\"No valid genome TXT files found in /content/. Check your input files.\")\n",
        "\n",
        "sex_codes = []\n",
        "individual_ids = []\n",
        "\n",
        "for filepath in genome_filepaths:\n",
        "    while True:\n",
        "        sex = input(f\"Enter sex for {filepath} (1 for male, 2 for female, 0 for unknown): \")\n",
        "        if sex in ['1', '2', '0']:\n",
        "            sex_codes.append(sex)\n",
        "            break\n",
        "        print(\"Invalid input. Please enter 1, 2, or 0.\")\n",
        "\n",
        "    name = input(f\"Enter name for {filepath} (to be displayed in .ind file): \")\n",
        "    individual_ids.append(name)\n",
        "\n",
        "# Output prefix\n",
        "output_base = input(\"Enter output base prefix: \").strip()\n",
        "\n",
        "if not output_base:\n",
        "    raise ValueError(\"Output base name cannot be empty.\")\n",
        "\n",
        "# Ask user if they want to trim using SNP list\n",
        "trim_with_snplist = input(\"Do you want to trim using a SNP list? (yes/no): \").strip().lower() == \"yes\"\n",
        "snplist_path = f\"/content/snplist/{dataset_name}snplist.snplist\" if trim_with_snplist else None\n",
        "\n",
        "# Add Swap Memory to Prevent Out-of-Memory Issues\n",
        "swap_file_path = \"/swapfile\"\n",
        "os.system(f\"fallocate -l 8G {swap_file_path}\")\n",
        "os.system(f\"chmod 600 {swap_file_path}\")\n",
        "os.system(f\"mkswap {swap_file_path}\")\n",
        "os.system(f\"swapon {swap_file_path}\")\n",
        "print(\"Swap memory created successfully.\")\n",
        "\n",
        "# Step 2: Convert Each Sample Separately and Trim if Required\n",
        "plink_file_prefixes = []\n",
        "for i, genome_filepath in enumerate(genome_filepaths):\n",
        "    sample_output_base = f\"{output_base}_{i+1}\"\n",
        "    plink_file_prefixes.append(sample_output_base)\n",
        "\n",
        "    print(f\"Processing {genome_filepath}...\")\n",
        "    subprocess.run([\n",
        "        \"/content/plink\", \"--23file\", genome_filepath, \"--list-23-indels\", \"--allow-no-sex\",\n",
        "        \"--make-bed\", \"--out\", sample_output_base\n",
        "    ], check=True)\n",
        "\n",
        "    if trim_with_snplist:\n",
        "        # Trim Using SNP List\n",
        "        trimmed_prefix = f\"{sample_output_base}_trimmed\"\n",
        "        subprocess.run([\n",
        "            \"/content/plink\", \"--bfile\", sample_output_base, \"--extract\", snplist_path,\n",
        "            \"--make-bed\", \"--out\", trimmed_prefix\n",
        "        ], check=True)\n",
        "        plink_file_prefixes[i] = trimmed_prefix  # Update prefix after trimming\n",
        "\n",
        "    #Define unimportant IDs so it doesn't give an error\n",
        "    family_ids = [\"FID\"] * len(genome_filepaths)  # Replace \"FID\" with actual family ID logic if needed\n",
        "    father_ids = [\"0\"] * len(genome_filepaths)   # Assuming unknown father (modify as needed)\n",
        "    mother_ids = [\"0\"] * len(genome_filepaths)   # Assuming unknown mother (modify as needed)\n",
        "    phenotype_values = [\"-9\"] * len(genome_filepaths)  # Default missing phenotype (-9 for PLINK)\n",
        "\n",
        "    #Define num_threads so it doesn't give an error\n",
        "    num_threads = 2\n",
        "\n",
        "    # Update .fam File\n",
        "    fam_file_path = f\"/content/{plink_file_prefixes[i]}.fam\"\n",
        "    new_fam_line = f\"{family_ids[i]} {individual_ids[i]} {father_ids[i]} {mother_ids[i]} {sex_codes[i]} {phenotype_values[i]}\\n\"\n",
        "\n",
        "    if os.path.exists(fam_file_path):\n",
        "        with open(fam_file_path, 'w') as fam_file:\n",
        "            fam_file.write(new_fam_line)\n",
        "        print(f\".fam file updated: {fam_file_path}\")\n",
        "    else:\n",
        "        raise FileNotFoundError(f\".fam file not found: {fam_file_path}\")\n",
        "\n",
        "# Step 3: Merge All PLINK Files\n",
        "print(\"Merging PLINK files...\")\n",
        "merge_command = f\"/content/plink --bfile {plink_file_prefixes[0]}\"\n",
        "for prefix in plink_file_prefixes[1:]:\n",
        "    merge_command += f\" --bmerge {prefix}\"\n",
        "\n",
        "merged_prefix = f\"{output_base}_merged\"\n",
        "merge_command += f\" --make-bed --allow-no-sex --out {merged_prefix}\"\n",
        "subprocess.run(merge_command, shell=True, check=True)\n",
        "print(f\"Merged PLINK file created: {merged_prefix}\")\n",
        "\n",
        "# Step 4: Trim Again After Merging if Requested\n",
        "if trim_with_snplist:\n",
        "    trimmed_merged_prefix = f\"{output_base}_merged_trimmed\"\n",
        "    subprocess.run([\n",
        "        \"/content/plink\", \"--bfile\", merged_prefix, \"--extract\", snplist_path,\n",
        "        \"--make-bed\", \"--out\", trimmed_merged_prefix\n",
        "    ], check=True)\n",
        "    merged_prefix = trimmed_merged_prefix  # Update prefix after trimming\n",
        "\n",
        "# Step 5: Convert to EIGENSTRAT Format\n",
        "bedfilepath = f\"/content/{merged_prefix}.bed\"\n",
        "bimfilepath = f\"/content/{merged_prefix}.bim\"\n",
        "famfilepath = f\"/content/{merged_prefix}.fam\"\n",
        "genotypeoutname = f\"{output_base}_converted.geno\"\n",
        "snpoutname = f\"{output_base}_converted.snp\"\n",
        "indivoutname = f\"{output_base}_converted.ind\"\n",
        "\n",
        "convertf_param_content = f\"\"\"\n",
        "genotypename: {bedfilepath}\n",
        "snpname: {bimfilepath}\n",
        "indivname: {famfilepath}\n",
        "genotypeoutname: {genotypeoutname}\n",
        "snpoutname: {snpoutname}\n",
        "indivoutname: {indivoutname}\n",
        "prodercheck: YES\n",
        "\"\"\"\n",
        "with open(\"/content/convertf_param.par\", \"w\") as param_file:\n",
        "    param_file.write(convertf_param_content)\n",
        "print(\"convertf parameter file created successfully.\")\n",
        "\n",
        "subprocess.run([\"convertf\", \"-p\", \"/content/convertf_param.par\"], check=True)\n",
        "\n",
        "# Step 6: Merge with Reference Dataset\n",
        "mergeit_param_content = f\"\"\"\n",
        "geno1: {geno1_path}\n",
        "snp1: {snp1_path}\n",
        "ind1: {ind1_path}\n",
        "geno2: {genotypeoutname}\n",
        "snp2: {snpoutname}\n",
        "ind2: {indivoutname}\n",
        "genooutfilename: {output_base}_merged.geno\n",
        "snpoutfilename: {output_base}_merged.snp\n",
        "indoutfilename: {output_base}_merged.ind\n",
        "testmismatch: NO\n",
        "numthreads: {num_threads}\n",
        "\"\"\"\n",
        "with open(\"/content/mergeit_param.par\", \"w\") as merge_param_file:\n",
        "    merge_param_file.write(mergeit_param_content)\n",
        "\n",
        "print(\"Running mergeit...\")\n",
        "!mergeit -p /content/mergeit_param.par\n",
        "\n",
        "# Verify Merging Results\n",
        "merged_files_exist = all(os.path.exists(f\"/content/{output_base}_merged.{ext}\") for ext in [\"geno\", \"snp\", \"ind\"])\n",
        "print(\"Merging completed successfully.\" if merged_files_exist else \"An error occurred during merging.\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uZG9UI3J36rd"
      },
      "source": [
        "# **[5] Download the merged files as a zip**"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hIYz130a4C1I"
      },
      "source": [
        "Just run the thing and give it the names (same one you gave it to the merged files)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 156
        },
        "id": "D1KBgV5EdcXu",
        "outputId": "e2d0500a-30e6-4805-f097-83d423563f21"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "import zipfile\n",
        "\n",
        "# Ask the user to specify the base filename for .ind, .geno, and .snp files\n",
        "base_filename = input(\"Enter the base filename for the merged files (without extension): \").strip()\n",
        "\n",
        "if not base_filename:\n",
        "    raise ValueError(\"Base filename cannot be empty.\")\n",
        "\n",
        "# Define the file extensions to include\n",
        "file_extensions = [\"geno\", \"snp\", \"ind\"]\n",
        "\n",
        "# Define excluded names\n",
        "excluded_names = [\"aadr_v62.0_HO_public\", \"aadr_v62.0_1240K_public\"]\n",
        "\n",
        "# Create a Zip File with Selected Files\n",
        "zip_filename = f\"{base_filename}_selected.zip\"\n",
        "zip_filepath = f\"/content/{zip_filename}\"\n",
        "\n",
        "print(f\"Creating ZIP archive: {zip_filename}...\")\n",
        "\n",
        "with zipfile.ZipFile(zip_filepath, 'w') as zipf:\n",
        "    for ext in file_extensions:\n",
        "        file_path = f\"/content/{base_filename}_merged.{ext}\"\n",
        "\n",
        "        # Skip files that match excluded names\n",
        "        if os.path.exists(file_path) and not any(excluded in file_path for excluded in excluded_names):\n",
        "            zipf.write(file_path, os.path.basename(file_path))\n",
        "        else:\n",
        "            print(f\"Skipping: {file_path} (not found or excluded)\")\n",
        "\n",
        "print(f\"ZIP archive created successfully: {zip_filepath}\")\n",
        "\n",
        "# Check if file exists\n",
        "if os.path.exists(zip_filepath):\n",
        "    print(f\"File found: {zip_filepath}\")\n",
        "    print(\"File size:\", round(os.path.getsize(zip_filepath) / (1024 * 1024), 2), \"MB\")\n",
        "\n",
        "    # Trigger download\n",
        "    print(\"\\nStarting download to your PC...\")\n",
        "    files.download(zip_filepath)\n",
        "    print(\"Check your browser for the download prompt!\")\n",
        "else:\n",
        "    print(\"Error: File not found at\", zip_filepath)\n",
        "    print(\"Please make sure:\")\n",
        "    print(\"1. The file exists in Colab's /content directory\")\n",
        "    print(\"2. You've run all previous cells that generate this file\")\n",
        "    print(\"3. The filename format matches '(BASEFILENAME)_merged'\")\n"
      ]
    }
  ],
  "metadata": {
    "accelerator": "TPU",
    "colab": {
      "gpuType": "V28",
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.13.2"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
